File: packages/oxlint-plugin-react-doctor/src/plugin/rules/state-and-effects/no-controlled-selection-focus-effect.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { findEnclosingFunction } from "../../utils/find-enclosing-function.js";
import { getCalleeName } from "../../utils/get-callee-name.js";
import { getEffectCallback } from "../../utils/get-effect-callback.js";
import { getStaticPropertyName } from "../../utils/get-static-property-name.js";
import { isFunctionLike } from "../../utils/is-function-like.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { isReactEffectHookCall } from "../../utils/is-react-effect-hook-call.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { stripParenExpression } from "../../utils/strip-paren-expression.js";

const MESSAGE =
  "This effect focuses an item whenever a controlled selection changes, so an external selection update can steal focus. Gate focus on user navigation or on the list opening.";

const IMPORTED_EFFECT_WRAPPER_NAMES = new Set([
  "useIsomorphicLayoutEffect",
  "useModernLayoutEffect",
]);

const CONTROLLED_SELECTION_NAME_PATTERN = /^(?:selected|selection)(?:index|id|key|item|value)?$/i;

const isRecognizedEffectCall = (
  call: EsTreeNodeOfType<"CallExpression">,
  context: RuleContext,
): boolean => {
  if (isReactEffectHookCall(call, context.scopes)) return true;
  const callee = stripParenExpression(call.callee);
  if (!isNodeOfType(callee, "Identifier") || !IMPORTED_EFFECT_WRAPPER_NAMES.has(callee.name)) {
    return false;
  }
  return context.scopes.symbolFor(callee)?.kind === "import";
};

const getControlledSelectionDependencies = (
  call: EsTreeNodeOfType<"CallExpression">,
  ownerFunction: EsTreeNode,
  context: RuleContext,
): ReadonlySet<number> => {
  const dependencyArgument = call.arguments[1];
  if (!dependencyArgument) return new Set();
  const dependencyArray = stripParenExpression(dependencyArgument);
  if (!isNodeOfType(dependencyArray, "ArrayExpression")) return new Set();
  const selectionSymbolIds = new Set<number>();
  for (const element of dependencyArray.elements) {
    if (!element) continue;
    const dependency = stripParenExpression(element);
    if (
      !isNodeOfType(dependency, "Identifier") ||
      !CONTROLLED_SELECTION_NAME_PATTERN.test(dependency.name)
    ) {
      continue;
    }
    const symbol = context.scopes.symbolFor(dependency);
    if (symbol?.kind === "parameter" && symbol.scope.node === ownerFunction) {
      selectionSymbolIds.add(symbol.id);
    }
  }
  return selectionSymbolIds;
};

const getSelectionRefSymbolId = (
  statement: EsTreeNode,
  selectionSymbolIds: ReadonlySet<number>,
  context: RuleContext,
): number | null => {
  if (!isNodeOfType(statement, "ExpressionStatement")) return null;
  const expression = stripParenExpression(statement.expression);
  if (!isNodeOfType(expression, "AssignmentExpression") || expression.operator !== "=") {
    return null;
  }
  const target = stripParenExpression(expression.left);
  const selection = stripParenExpression(expression.right);
  if (
    !isNodeOfType(target, "MemberExpression") ||
    getStaticPropertyName(target) !== "current" ||
    !isNodeOfType(selection, "Identifier")
  ) {
    return null;
  }
  const selectionSymbol = context.scopes.symbolFor(selection);
  if (!selectionSymbol || !selectionSymbolIds.has(selectionSymbol.id)) return null;
  const reference = stripParenExpression(target.object);
  if (!isNodeOfType(reference, "Identifier")) return null;
  const referenceSymbol = context.scopes.symbolFor(reference);
  if (!referenceSymbol) return null;
  return referenceSymbol.id;
};

const isFocusCallUsingReference = (
  statement: EsTreeNode,
  referenceSymbolId: number,
  context: RuleContext,
): boolean => {
  if (!isNodeOfType(statement, "ExpressionStatement")) return false;
  const expression = stripParenExpression(statement.expression);
  if (!isNodeOfType(expression, "CallExpression")) return false;
  const calleeName = getCalleeName(expression);
  if (!calleeName?.startsWith("focus")) return false;
  return expression.arguments.some((argument) => {
    const value = stripParenExpression(argument);
    return (
      isNodeOfType(value, "Identifier") && context.scopes.symbolFor(value)?.id === referenceSymbolId
    );
  });
};

const findControlledSelectionFocus = (
  call: EsTreeNodeOfType<"CallExpression">,
  context: RuleContext,
): boolean => {
  const ownerFunction = findEnclosingFunction(call);
  const callback = getEffectCallback(call, context.scopes);
  if (
    !ownerFunction ||
    !isFunctionLike(callback) ||
    !isNodeOfType(callback.body, "BlockStatement")
  ) {
    return false;
  }
  const selectionSymbolIds = getControlledSelectionDependencies(call, ownerFunction, context);
  if (selectionSymbolIds.size === 0) return false;
  const selectionRefSymbolIds = new Set<number>();
  for (const statement of callback.body.body) {
    const selectionRefSymbolId = getSelectionRefSymbolId(statement, selectionSymbolIds, context);
    if (selectionRefSymbolId !== null) selectionRefSymbolIds.add(selectionRefSymbolId);
    for (const referenceSymbolId of selectionRefSymbolIds) {
      if (isFocusCallUsingReference(statement, referenceSymbolId, context)) return true;
    }
  }
  return false;
};

export const noControlledSelectionFocusEffect = defineRule({
  id: "no-controlled-selection-focus-effect",
  title: "Controlled selection change steals focus",
  severity: "warn",
  category: "Accessibility",
  tags: ["react-jsx-only"],
  defaultEnabled: false,
  recommendation:
    "Move selection-driven focus into the keyboard or pointer navigation path, or prove that the list just opened before focusing the selected item.",
  create: (context: RuleContext) => ({
    CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
      if (!isRecognizedEffectCall(node, context)) return;
      if (findControlledSelectionFocus(node, context)) context.report({ node, message: MESSAGE });
    },
  }),
});

File: native/oxlint/rules/no-controlled-selection-focus-effect.rs
use oxc_ast::{
    AstKind,
    ast::{Argument, ArrayExpressionElement, Expression, Statement},
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_semantic::{NodeId, SymbolId};
use oxc_span::GetSpan;
use oxc_syntax::operator::AssignmentOperator;
use rustc_hash::FxHashSet;

use crate::{AstNode, context::LintContext, rule::Rule};

const EFFECT_HOOK_NAMES: [&str; 2] = ["useEffect", "useLayoutEffect"];
const IMPORTED_EFFECT_WRAPPER_NAMES: [&str; 2] =
    ["useIsomorphicLayoutEffect", "useModernLayoutEffect"];
const MESSAGE: &str = "This effect focuses an item whenever a controlled selection changes, so an external selection update can steal focus. Gate focus on user navigation or on the list opening.";

#[derive(Debug, Default, Clone)]
pub struct NoControlledSelectionFocusEffect;

declare_oxc_lint!(
    /// Warns when controlled selection changes drive focus from an effect.
    NoControlledSelectionFocusEffect,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Warns when controlled selection changes steal focus.",
);

impl Rule for NoControlledSelectionFocusEffect {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::CallExpression(effect_call) = node.kind() else {
            return;
        };
        if !is_controlled_selection_effect_call(effect_call, ctx) {
            return;
        }
        let Some(owner_function_id) = local_callback_nearest_function_id(node.id(), ctx) else {
            return;
        };
        let selection_symbol_ids =
            controlled_selection_dependency_symbols(effect_call, owner_function_id, ctx);
        if selection_symbol_ids.is_empty() {
            return;
        }
        let Some(callback_expression) = effect_call
            .arguments
            .first()
            .and_then(Argument::as_expression)
        else {
            return;
        };
        let Some(callback_id) = exact_local_function_id_including_generators(
            callback_expression,
            ctx,
            &mut Vec::new(),
            &mut LocalFunctionResolutionCache::default(),
        ) else {
            return;
        };
        let callback = ctx.nodes().get_node(callback_id);
        let statements = match callback.kind() {
            AstKind::Function(function) => function.body.as_ref().map(|body| &body.statements),
            AstKind::ArrowFunctionExpression(function) if function.get_expression().is_none() => {
                function
                    .body
                    .as_function_body()
                    .map(|body| &body.statements)
            }
            _ => None,
        };
        let Some(statements) = statements else {
            return;
        };
        let mut selection_ref_symbol_ids = FxHashSet::default();
        for statement in statements {
            if let Some(reference_symbol_id) =
                selection_ref_symbol_id(statement, &selection_symbol_ids, ctx)
            {
                selection_ref_symbol_ids.insert(reference_symbol_id);
            }
            if selection_ref_symbol_ids.iter().any(|reference_symbol_id| {
                focus_call_uses_reference(statement, *reference_symbol_id, ctx)
            }) {
                ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(effect_call.span));
                return;
            }
        }
    }
}

fn is_controlled_selection_effect_call<'a>(
    call_expression: &oxc_ast::ast::CallExpression<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    if is_react_hook_call(call_expression, &EFFECT_HOOK_NAMES, ctx) {
        return true;
    }
    let Expression::Identifier(identifier) = call_expression.callee.get_inner_expression() else {
        return false;
    };
    if !IMPORTED_EFFECT_WRAPPER_NAMES.contains(&identifier.name.as_str()) {
        return false;
    }
    let Some(symbol_id) = ctx
        .scoping()
        .get_reference(identifier.reference_id())
        .symbol_id()
    else {
        return false;
    };
    matches!(
        ctx.symbol_declaration(symbol_id).kind(),
        AstKind::ImportSpecifier(_)
            | AstKind::ImportDefaultSpecifier(_)
            | AstKind::ImportNamespaceSpecifier(_)
    )
}

fn controlled_selection_dependency_symbols(
    call_expression: &oxc_ast::ast::CallExpression<'_>,
    owner_function_id: NodeId,
    ctx: &LintContext<'_>,
) -> FxHashSet<SymbolId> {
    let Some(Expression::ArrayExpression(dependencies)) = call_expression
        .arguments
        .get(1)
        .and_then(Argument::as_expression)
        .map(Expression::get_inner_expression)
    else {
        return FxHashSet::default();
    };
    dependencies
        .elements
        .iter()
        .filter_map(ArrayExpressionElement::as_expression)
        .filter_map(|dependency| {
            let Expression::Identifier(identifier) = dependency.get_inner_expression() else {
                return None;
            };
            if !is_controlled_selection_name(identifier.name.as_str()) {
                return None;
            }
            let symbol_id = ctx
                .scoping()
                .get_reference(identifier.reference_id())
                .symbol_id()?;
            let declaration = ctx.symbol_declaration(symbol_id);
            if !matches!(declaration.kind(), AstKind::FormalParameter(_)) {
                return None;
            }
            ctx.nodes()
                .ancestors(declaration.id())
                .find(|ancestor| {
                    matches!(
                        ancestor.kind(),
                        AstKind::Function(_) | AstKind::ArrowFunctionExpression(_)
                    )
                })
                .is_some_and(|function| function.id() == owner_function_id)
                .then_some(symbol_id)
        })
        .collect()
}

fn is_controlled_selection_name(name: &str) -> bool {
    let lowercase_name = name.to_ascii_lowercase();
    ["selected", "selection"].iter().any(|prefix| {
        lowercase_name.strip_prefix(prefix).is_some_and(|suffix| {
            suffix.is_empty() || ["index", "id", "key", "item", "value"].contains(&suffix)
        })
    })
}

fn selection_ref_symbol_id(
    statement: &Statement<'_>,
    selection_symbol_ids: &FxHashSet<SymbolId>,
    ctx: &LintContext<'_>,
) -> Option<SymbolId> {
    let Statement::ExpressionStatement(statement) = statement else {
        return None;
    };
    let Expression::AssignmentExpression(assignment) = statement.expression.get_inner_expression()
    else {
        return None;
    };
    if assignment.operator != AssignmentOperator::Assign {
        return None;
    }
    let target = assignment.left.as_member_expression()?;
    if static_member_expression_property_name(target) != Some("current") {
        return None;
    }
    let Expression::Identifier(selection) = assignment.right.get_inner_expression() else {
        return None;
    };
    let selection_symbol_id = ctx
        .scoping()
        .get_reference(selection.reference_id())
        .symbol_id()?;
    if !selection_symbol_ids.contains(&selection_symbol_id) {
        return None;
    }
    let Expression::Identifier(reference) = target.object().get_inner_expression() else {
        return None;
    };
    ctx.scoping()
        .get_reference(reference.reference_id())
        .symbol_id()
}

fn focus_call_uses_reference(
    statement: &Statement<'_>,
    reference_symbol_id: SymbolId,
    ctx: &LintContext<'_>,
) -> bool {
    let Statement::ExpressionStatement(statement) = statement else {
        return false;
    };
    let Expression::CallExpression(call_expression) = statement.expression.get_inner_expression()
    else {
        return false;
    };
    let callee_name = match &call_expression.callee {
        Expression::Identifier(identifier) => Some(identifier.name.as_str()),
        expression => expression
            .as_member_expression()
            .and_then(member_expression_identifier_property_name),
    };
    if !callee_name.is_some_and(|name| name.starts_with("focus")) {
        return false;
    }
    call_expression.arguments.iter().any(|argument| {
        let Some(Expression::Identifier(identifier)) = argument
            .as_expression()
            .map(Expression::get_inner_expression)
        else {
            return false;
        };
        ctx.scoping()
            .get_reference(identifier.reference_id())
            .symbol_id()
            == Some(reference_symbol_id)
    })
}
